packages/grid/_modules_/grid/components/cell/GridTreeDataGroupingCell.tsx

```diff
@@ -0,0 +1,170 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { unstable_composeClasses as composeClasses } from '@mui/core';
+import IconButton from '@mui/material/IconButton';
+import Box from '@mui/material/Box';
+import { useGridRootProps } from '../../hooks/utils/useGridRootProps';
+import { useGridApiContext } from '../../hooks/utils/useGridApiContext';
+import { GridRenderCellParams } from '../../models/params/gridCellParams';
+import { isNavigationKey, isSpaceKey } from '../../utils/keyboardUtils';
+import { GridEvents } from '../../constants';
+import { getDataGridUtilityClass } from '../../gridClasses';
+import { GridComponentProps } from '../../GridComponentProps';
+
+type OwnerState = { classes: GridComponentProps['classes'] };
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    root: ['treeDataGroupingCell'],
+    toggle: ['treeDataGroupingCellToggle'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+export interface GridTreeDataGroupingCellValue {
+  label: string;
+  filteredDescendantCount: number;
+  depth: number;
+  expanded: boolean;
+}
+
+const GridTreeDataGroupingCell = (props: GridRenderCellParams<GridTreeDataGroupingCellValue>) => {
+  const { id, field, value } = props;
+
+  const rootProps = useGridRootProps();
+  const apiRef = useGridApiContext();
+  const ownerState: OwnerState = { classes: rootProps.classes };
+  const classes = useUtilityClasses(ownerState);
+
+  const Icon = value.expanded
+    ? rootProps.components.TreeDataCollapseIcon
+    : rootProps.components.TreeDataExpandIcon;
+
+  const handleKeyDown = (event) => {
+    if (isSpaceKey(event.key)) {
+      event.stopPropagation();
+    }
+    if (isNavigationKey(event.key) && !event.shiftKey) {
+      apiRef.current.publishEvent(GridEvents.cellNavigationKeyDown, props, event);
+    }
+  };
+
+  const handleClick = (event) => {
+    apiRef.current.unstable_setRowExpansion(id, !value.expanded);
+    apiRef.current.setCellFocus(id, field);
+    event.stopPropagation();
```

Comment:
Stopping the propagation causes side effects. See https://github.com/mui-org/material-ui-x/issues/3158#issuecomment-966644362